In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df=pd.read_csv("C:/Users/knowl/OneDrive/Desktop/Titanic Competition/train.csv")

In [3]:
df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [4]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

We have many missing values in Age column, we will fill the missing values with Median age. 

In [6]:
df['Age'] = df['Age'].fillna(df['Age'].median())
df['Embarked'] = df['Embarked'].fillna(df['Embarked'].mode()[0])


In [7]:
df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male  28.0      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [8]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  891.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.361582    0.523008   
std     257.353842    0.486592    0.836071   13.019697    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   22.000000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   35.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

I have handeled the missing value in the Age column. 

Feature Engineering- In this step, I created new features from existing columns to help machine learning model capture important patterns.

Features Created:
1- FamilySize

Combines SibSp (siblings/spouse) and Parch (parents/children) to represent total family size.

Formula: FamilySize = SibSp + Parch + 1 (the passenger is included).

2- IsAlone

A binary feature that indicates if a passenger is traveling alone.

If FamilySize == 1, then IsAlone = 1; otherwise, 0.

3- Title

Extracted from the Name column (e.g., Mr., Miss., Dr.).

Helps capture social status, gender, and age-related cues.

Rare titles are grouped under 'Rare' and similar titles are merged (e.g., Mlle → Miss).

These engineered features can improve model performance by adding new meaningful insights not explicitly visible in the original data.

In [11]:
df['FamilySize'] = df['SibSp'] + df['Parch'] + 1


In [12]:
df['IsAlone'] = 0
df.loc[df['FamilySize'] == 1, 'IsAlone'] = 1


In [13]:
df['Title'] = df['Name'].str.extract(r' ([A-Za-z]+)\.', expand=False)


In [14]:
df['Title'] = df['Title'].replace([
    'Lady', 'Countess', 'Capt', 'Col', 'Don', 'Dr', 'Major', 'Rev', 
    'Sir', 'Jonkheer', 'Dona'
], 'Rare')
df['Title'] = df['Title'].replace({'Mlle': 'Miss', 'Ms': 'Miss', 'Mme': 'Mrs'})


In [15]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  FamilySize  IsAlone Title  
0      0         A/5 21171   7.2500   NaN        S           2        0    Mr  
1      0          PC 17599  71.2833   C85        C           2        0   Mrs  
2      0  STON/O2. 3101282   7.9250   NaN        S           1        1  Miss  
3      0            113803  53.1000  C123        S           2        0   Mrs  
4      0            373450   8.0500   NaN        S           1        1    Mr

Machine learning models require all input features to be numeric. However, our dataset still contains categorical variables like Sex, Embarked, and Title. To make these usable by the model, we convert them into numerical form using encoding techniques:

Label Encoding:

Converted the Sex column into binary values:

'male' → 0

'female' → 1

One-Hot Encoding:

For Embarked and Title (which have multiple categories), we created new binary columns for each category using one-hot encoding.

We set drop_first=True to avoid multicollinearity (i.e., avoid dummy variable trap).

In [17]:
# Convert 'Sex' column: male = 0, female = 1
df['Sex'] = df['Sex'].map({'male': 0, 'female': 1}).astype(int)

# One-hot encode 'Embarked' (drop_first=True to avoid multicollinearity)
df = pd.get_dummies(df, columns=['Embarked'], drop_first=True)

# One-hot encode 'Title' (drop_first=True to avoid multicollinearity)
df = pd.get_dummies(df, columns=['Title'], drop_first=True)


I will now drop unecessary columns which my model does not need.

In [19]:
df.drop(['Name', 'Ticket', 'Cabin', 'PassengerId'], axis=1, inplace=True)


In [20]:
df.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  FamilySize  IsAlone  \
0         0       3    0  22.0      1      0   7.2500           2        0   
1         1       1    1  38.0      1      0  71.2833           2        0   
2         1       3    1  26.0      0      0   7.9250           1        1   
3         1       1    1  35.0      1      0  53.1000           2        0   
4         0       3    0  35.0      0      0   8.0500           1        1   

   Embarked_Q  Embarked_S  Title_Miss  Title_Mr  Title_Mrs  Title_Rare  
0       False        True       False      True      False       False  
1       False       False       False     False       True       False  
2       False        True        True     False      False       False  
3       False        True       False     False       True       False  
4       False        True       False      True      False       False

In [21]:
df = df.astype(int)


In [22]:
df.head()

Survived  Pclass  Sex  Age  SibSp  Parch  Fare  FamilySize  IsAlone  \
0         0       3    0   22      1      0     7           2        0   
1         1       1    1   38      1      0    71           2        0   
2         1       3    1   26      0      0     7           1        1   
3         1       1    1   35      1      0    53           2        0   
4         0       3    0   35      0      0     8           1        1   

   Embarked_Q  Embarked_S  Title_Miss  Title_Mr  Title_Mrs  Title_Rare  
0           0           1           0         1          0           0  
1           0           0           0         0          1           0  
2           0           1           1         0          0           0  
3           0           1           0         0          1           0  
4           0           1           0         1          0           0

 In this step, I train a Logistic Regression model to predict passenger survival. Logistic Regression is a simple and effective classification algorithm suitable for binary outcomes (like Survived = 0 or 1).

-What I Did:
Split the Data

The dataset was split into:

X: All input features

y: Target variable (Survived)

Then I split the data into:

Training Set (80%) – to train the model

Validation Set (20%) – to evaluate model accuracy

Train the Model



Split features & target

In [25]:
# Split features and target
X = df.drop('Survived', axis=1)  # Features
y = df['Survived']               # Target


I will use Logistic Regression for binary classification

In [27]:
from sklearn.model_selection import train_test_split

# Features and target
X = df.drop('Survived', axis=1)
y = df['Survived']

# Split into train (80%) and validation (20%)
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)


In [28]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)


LogisticRegression(max_iter=1000)

In [29]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, classification_report

# Predict on validation set
y_pred = model.predict(X_val)

# Individual metrics
print("Accuracy:", accuracy_score(y_val, y_pred))
print("Precision:", precision_score(y_val, y_pred))
print("Recall:", recall_score(y_val, y_pred))
print("F1 Score:", f1_score(y_val, y_pred))

# Full classification report
print("\nClassification Report:\n", classification_report(y_val, y_pred))


Accuracy: 0.8044692737430168
Precision: 0.7671232876712328
Recall: 0.7567567567567568
F1 Score: 0.7619047619047619

Classification Report:
               precision    recall  f1-score   support

           0       0.83      0.84      0.83       105
           1       0.77      0.76      0.76        74

    accuracy                           0.80       179
   macro avg       0.80      0.80      0.80       179
weighted avg       0.80      0.80      0.80       179



### 📊 Logistic Regression Evaluation Summary

I evaluated my logistic regression model on the validation set and achieved an **accuracy of 80.4%**, with a **precision of 76.7%**, **recall of 75.7%**, and an **F1-score of 76.1%**. These results indicate that my model performs well in predicting survival outcomes.

Class 0 (not survived) had slightly better performance than class 1 (survived), which is expected given the nature of the dataset. Overall, the model demonstrates a good balance between precision and recall.

I consider this a strong baseline for the Titanic classification problem. In the future, I can explore more advanced models such as Random Forest or XGBoost, and experiment with hyperparameter tuning to further improve the prediction accuracy.


## 📦 Predicting on Test Dataset 


In [32]:
df=pd.read_csv("C:/Users/knowl/OneDrive/Desktop/Titanic Competition/test.csv")

In [33]:
df.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

In [34]:
df.describe()

PassengerId      Pclass         Age       SibSp       Parch        Fare
count   418.000000  418.000000  332.000000  418.000000  418.000000  417.000000
mean   1100.500000    2.265550   30.272590    0.447368    0.392344   35.627188
std     120.810458    0.841838   14.181209    0.896760    0.981429   55.907576
min     892.000000    1.000000    0.170000    0.000000    0.000000    0.000000
25%     996.250000    1.000000   21.000000    0.000000    0.000000    7.895800
50%    1100.500000    3.000000   27.000000    0.000000    0.000000   14.454200
75%    1204.750000    3.000000   39.000000    1.000000    0.000000   31.500000
max    1309.000000    3.000000   76.000000    8.000000    9.000000  512.329200

Filling missing age and fair with median

In [36]:
df['Age'] = df['Age'].fillna(df['Age'].median())
df['Fare'] = df['Fare'].fillna(df['Fare'].median())

### 🏗️ Feature Engineering: Family-Based Insights

In this step, I created two new features from the existing `SibSp` and `Parch` columns to better capture the social structure of each passenger:

- **`FamilySize`**: Calculated as the total number of family members a passenger was traveling with, including themselves.  
  `FamilySize = SibSp + Parch + 1`

- **`IsAlone`**: A binary feature indicating whether a passenger was traveling alone.  
  If `FamilySize == 1`, then `IsAlone = 1`; otherwise, `IsAlone = 0`.

These engineered features help the model capture survival patterns related to social grouping — for example, passengers traveling alone may have had different survival odds than those with family.


In [38]:
df['FamilySize'] = df['SibSp'] + df['Parch'] + 1
df['IsAlone'] = 0
df.loc[df['FamilySize'] == 1, 'IsAlone'] = 1

Extracting title from Name

In [40]:
df['Title'] = df['Name'].str.extract(r' ([A-Za-z]+)\.', expand=False)
df['Title'] = df['Title'].replace(['Lady', 'Countess', 'Capt', 'Col', 'Don', 'Dr', 'Major', 'Rev', 'Sir', 'Jonkheer', 'Dona'], 'Rare')
df['Title'] = df['Title'].replace({'Mlle': 'Miss', 'Ms': 'Miss', 'Mme': 'Mrs'})

### 🏷️ Feature Engineering: Title Extraction from Name

In this step, I extracted the Title (such as *Mr.*, *Miss.*, *Dr.*) from each passenger's `Name` column using a regular expression. Titles can provide meaningful information about a passenger’s age, gender, and social status — all of which may influence their chances of survival.

#### 🔧 What I did:
- Extracted titles using the pattern r' ([A-Za-z]+)\.' from the `Name` column.
- Grouped less common or rare titles such as `Capt`, `Sir`, `Countess`, and `Rev` into a single category: `'Rare'`.
- Replaced alternate spellings and equivalents:
  - 'Mlle', 'Ms' → 'Miss'
  - 'Mme' → 'Mrs'

This `Title` feature adds social and demographic context, helping improve the model’s predictive power.


Encoding Categorical variable

In [43]:
df['Sex'] = df['Sex'].map({'male': 0, 'female': 1}).astype(int)
df = pd.get_dummies(df, columns=['Embarked'], drop_first=True)
df = pd.get_dummies(df, columns=['Title'], drop_first=True)

In [44]:
# Align columns with X (training data features)
X_test = df.drop(['PassengerId', 'Name', 'Ticket', 'Cabin'], axis=1)
X_test = X_test.reindex(columns=X.columns, fill_value=0)

In [45]:
predictions = model.predict(X_test)

In [46]:
print(predictions)


[0 1 0 0 1 0 1 0 1 0 0 0 1 0 1 1 0 0 1 1 0 1 1 1 1 0 1 0 0 0 0 0 1 1 0 0 1
 1 0 0 0 0 0 1 1 0 0 0 1 1 0 0 1 1 0 0 0 0 0 1 0 0 0 1 1 1 1 0 0 1 1 0 1 0
 1 1 0 1 0 1 1 0 0 0 0 0 1 1 1 1 1 0 1 0 0 0 1 0 1 0 1 0 0 0 1 0 0 0 0 0 0
 1 1 1 1 0 0 1 0 1 1 0 1 0 0 1 0 1 0 0 0 1 1 0 0 0 0 0 1 0 0 1 0 0 0 0 0 0
 0 0 1 0 0 1 0 0 1 1 0 1 1 1 1 0 0 1 0 0 1 1 0 0 0 0 0 1 1 0 1 1 0 0 1 0 1
 0 1 0 0 0 0 0 0 0 1 0 1 1 0 1 1 1 1 1 0 0 1 0 1 0 0 0 0 1 0 0 1 0 1 0 1 0
 1 0 1 1 0 1 0 0 0 1 0 0 0 0 0 0 1 1 1 1 0 0 0 0 1 0 1 1 1 0 1 0 0 0 0 0 1
 0 0 0 1 1 0 0 0 0 1 0 0 0 1 1 0 1 0 0 0 0 1 1 1 1 1 0 0 0 0 0 0 1 0 0 0 0
 1 0 0 0 0 0 0 0 1 1 0 1 0 1 0 0 0 1 1 1 0 0 0 0 0 0 0 0 1 0 1 0 0 0 1 0 0
 1 0 0 0 0 0 1 0 0 0 1 0 1 0 1 0 1 1 0 0 0 1 0 1 0 0 1 0 1 1 0 1 0 0 1 1 0
 0 1 0 0 1 1 0 0 0 0 0 0 1 1 0 1 0 0 0 0 0 1 0 0 0 1 0 1 0 0 1 0 1 0 0 0 0
 0 1 1 1 1 1 0 1 0 0 0]


### ✅ Project Summary

In this Titanic machine learning project, I successfully:
- Explored and cleaned real-world passenger data
- Engineered features like `Title`, `FamilySize`, and `IsAlone`
- Converted categorical variables into numerical format using label and one-hot encoding
- Built and evaluated a Logistic Regression model
- Achieved **80.4% validation accuracy** with balanced precision and recall
- Generated predictions on the Kaggle test set and exported them to `submission.csv`

This project helped me apply core machine learning concepts — from preprocessing to model deployment — using Python, Pandas, and Scikit-learn.
